onnxruntime/core/framework/utils.cc

```diff
@@ -161,6 +161,14 @@ static Status BatchOrCopyMLValue(const SessionState& session_state,
     return Status::OK();
   }
 
+#ifdef USE_DML
+  // The DML EP supports binding external allocations directly, even if the memory types don't match, as long as they are on the same D3D12 device
+  if (copy_info.source_device.Type() == OrtDevice::GPU && copy_info.target_device.Type() == OrtDevice::GPU && copy_info.source_device.Id() == copy_info.target_device.Id()) {
+    target_mlvalue = source_mlvalue;
+    return Status::OK();
+  }
+#endif
```

Comment:
Is this really true, that if I have a combined CUDA+DML build of ORT, I can just treat `MemoryType DEFAULT` the same as `MemoryType CUDA_PINNED`, without regard for the memory type? Or is that a case we don't have to worry about because an earlier EP comparison would prevent this function from being called in the first place with CUDA GPU and DML GPU resources?

Reply:
I'm looking at the existing memory types and wondering in what case DML would see something besides DEFAULT?

```
  struct MemType {
    // Pre-defined memory types.
    static const MemoryType DEFAULT = 0;
    static const MemoryType CUDA_PINNED = 1;
    static const MemoryType HIP_PINNED = 2;
    static const MemoryType CANN_PINNED = 3;
  };
```

Reply:
This PR adds a DML_EXTERNAL type to differentiate between memory that has been allocated internally and what the user allocated.

There are so many things that would need to change for a DML+CUDA build to work that it's not something that we should worry about, at least not yet.

Reply:
Certainly not DML+CUDA at the same time, but Jeff and I have used DML+CUDA builds before (only running DML or CUDA per a single session) to run cases and compare the outputs (easier than building twice and switching between two completely separate builds). So, I just want to be sure we're not adding *more* bugs for us to worry about in the future...

Reply:
I don't think ORT should ever be built that way though, no? It happens to work in most scenarios, but I doubt IO binding is working in its current state when building with both CUDA and DML support at the same time.

This existing logic `if (copy_info.source_device == copy_info.target_device)` isn't able to differentiate between a DML or CUDA GPU, so it would presumably already fail. Both DML and CUDA use the DEFAULT memory type and the GPU device type.

I'll add a guard to at least make this new condition only be true when dealing with DML devices, but I don't think it really changes anything.